newrelicutil.go

```diff
@@ -5,42 +5,42 @@ import (
 	"context"
 	"net/http"
 
-	"github.com/newrelic/go-agent"
+	"github.com/newrelic/go-agent/v3/newrelic"
 )
 
 type key int
 
 const (
-	transaction key = iota
-	segment
+	segment key = iota
 	externalSegment
 	datastoreSegment
 )
 
 // Transaction returns the New Relic Transaction object from context.
-func Transaction(ctx context.Context) newrelic.Transaction {
-	if txn, ok := ctx.Value(transaction).(newrelic.Transaction); ok {
-		return txn
-	}
-	return nil
+// Returns nil if the context does not contain an object.
+func Transaction(ctx context.Context) *newrelic.Transaction {
+	return newrelic.FromContext(ctx)
 }
 
 // WithTransaction puts the New Relic Transaction object to the given context
 // and returns the new context.
-func WithTransaction(ctx context.Context, txn newrelic.Transaction) context.Context {
-	return context.WithValue(ctx, transaction, txn)
+func WithTransaction(ctx context.Context, txn *newrelic.Transaction) context.Context {
+	return newrelic.NewContext(ctx, txn)
```

Comment:
I would also leave the old logic:
```go
func Transaction(ctx context.Context) newrelic.Transaction {
        if txn := newrelic.FromContext(ctx); txn != nil {
            return txn
        }

        // for potential old users -- we will delete this after migration
	if txn, ok := ctx.Value(transaction).(newrelic.Transaction); ok {
		return txn
	}

	return nil
}

func WithTransaction(ctx context.Context, txn newrelic.Transaction) context.Context {
        ctx := newrelic.NewContext(ctx, txn)

        // for potential old users -- we will delete this after migration
	return context.WithValue(ctx, transaction, txn)
}
```

Reply:
This is a potential use-case. I made up this. But it worth checking.
```
service -- uses utils@v1:  newrelicutils.WithTransaction(ctx, tx) sets new context key
|
package -- uses utils@v2: newrelicutils.Transaction(ctx) reads old context key
```

Reply:
in any case utils@v1.Transaction will return nil as it won't be able to assert to Transaction interface

Reply:
After a discussion, we decided that it doesn't make sense to try to solve compatibility problem between newrelic agent v2 and v3 - it's almost impossible. It's better to give an opportunity to fallback to newrelicutil with agent v2, therefore newrelicutil with the agent v3 will get the tag v2.0.0.